python/lsst/pipe/base/pipeline_graph/expressions.py

```diff
@@ -0,0 +1,274 @@
+# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
+# respectively.  If you choose the GPL option then the following text applies
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Expressions that resolve to subsets of pipelines.
+
+See :ref:`pipeline-graph-subset-expressions`.
+"""
+
+from __future__ import annotations
+
+__all__ = (
+    "DirectionNode",
+    "IdentifierNode",
+    "IntersectionNode",
+    "Node",
+    "NotNode",
+    "UnionNode",
+    "parse",
+)
+
+import dataclasses
+import functools
+from typing import TYPE_CHECKING, Any, Literal, TypeAlias
+
+from lsst.daf.butler.registry.queries.expressions.parser.ply import lex, yacc
+
+from ._exceptions import InvalidExpressionError
+
+if TYPE_CHECKING:
+    from lsst.daf.butler.registry.queries.expressions.parser.parserLex import LexToken
+    from lsst.daf.butler.registry.queries.expressions.parser.parserYacc import YaccProduction
+
+
+class _ParserLex:
+    @classmethod
+    def make_lexer(cls) -> Any:  # unspecified PLY type.
+        return lex.lex(object=cls())
+
+    tokens = (
+        "IDENTIFIER",
+        "LPAREN",
+        "RPAREN",
+        "NOT",
+        "UNION",
+        "INTERSECTION",
+        "LT",
+        "LE",
+        "GT",
+        "GE",
+    )
+
+    t_LPAREN = r"\("
+    t_RPAREN = r"\)"
+    t_NOT = "~"
+    t_UNION = r"\|"
+    t_INTERSECTION = "&"
+    t_LT = "<"
+    t_LE = "<="
+    t_GT = ">"
+    t_GE = ">="
+
+    # Identifiers are alphanumeric, and may have a T:, D:, or S: prefix.
+    def t_IDENTIFIER(self, t: LexToken) -> LexToken:
+        r"""([TDS]:)?\w+"""
+        t.type = "IDENTIFIER"
+        return t
+
+    # Ignore spaces and tables.
+    t_ignore = " \t"
+
+    def t_error(self, t: LexToken) -> LexToken:
+        raise InvalidExpressionError(
+            f"invalid token in expression near character {t.lexer.lexpos}: {t.value[0]!r}"
+        )
+
+
+class _ParserYacc:
+    def __init__(self) -> None:
+        self.parser = self._parser_factory()
+
+    @staticmethod
+    @functools.cache
+    def _parser_factory() -> Any:  # unspecified PLY type.
+        return yacc.yacc(module=_ParserYacc, write_tables=False, debug=False)
+
+    def parse(self, input: str) -> Node:
+        """Parse input expression ad return parsed tree object.
```

Comment:
ad -> and

Reply:
Looks like I cargo-culted some of the rest of this docstring from Andy's daf_butler docstrings; I've droppped the second paragraph below, too.